driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2349,6 +2334,255 @@ public void run() {
+                return address + " " + eventType;
+            }
+
+        }
+
+        private class NodeRefreshRequestDeliveryCallback implements EventDebouncer.DeliveryCallback<NodeRefreshRequest> {
+
+            @Override
+            public ListenableFuture<?> deliver(List<NodeRefreshRequest> events) {
+                Map<InetSocketAddress, HostEvent> hosts = new HashMap<InetSocketAddress, HostEvent>();
+                // only keep the last event for each host
+                for (NodeRefreshRequest req : events) {
+                    hosts.put(req.address, req.eventType);
+                }
+                List<ListenableFuture<?>> futures = new ArrayList<ListenableFuture<?>>(hosts.size());
+                for (final Entry<InetSocketAddress, HostEvent> entry : hosts.entrySet()) {
+                    InetSocketAddress address = entry.getKey();
+                    HostEvent eventType = entry.getValue();
+                    switch (eventType) {
+                        case UP:
+                            Host upHost = metadata.getHost(address);
+                            if (upHost == null) {
+                                upHost = metadata.add(address);
+                                // If upHost is still null, it means we didn't know about it the line before but
+                                // got beaten at adding it to the metadata by another thread. In that case, it's
+                                // fine to let the other thread win and ignore the notification here
+                                if (upHost == null)
+                                    continue;
+                                futures.add(schedule(hostAdded(upHost)));
+                            } else {
+                                futures.add(schedule(hostUp(upHost)));
+                            }
+                            break;
+                        case ADDED:
+                            Host newHost = metadata.add(address);
+                            if (newHost != null)
+                                futures.add(schedule(hostAdded(newHost)));
+                            break;
+                        case DOWN:
+                            // Note that there is a slight risk we can receive the event late and thus
+                            // mark the host down even though we already had reconnected successfully.
+                            // But it is unlikely, and don't have too much consequence since we'll try reconnecting
+                            // right away, so we favor the detection to make the Host.isUp method more reliable.
+                            Host downHost = metadata.getHost(address);
+                            if (downHost != null)
+                                futures.add(execute(hostDown(downHost)));
+                            break;
+                        case REMOVED:
+                            Host removedHost = metadata.getHost(address);
```

Comment:
Previously this did:

```java
removeHost(metadata.getHost(tpAddr), false);
```

Now it will do `hostRemoved` which refreshesNodeInfo first, the node will be gone from peers, so it skips calling onRemove altogether since it is no longer in peers.   I think this is a straightforward fix so I'll commit it, but feel free to back out if you disagree.